dod-advana/gamechanger-data
# Improve parser efficiency: reference extraction

This PR is part of the effort to enhance parser efficiency and specifically focuses on reference extraction.

## Summary of Changes:

[ref_utils.py](https://github.com/dod-advana/gamechanger-data/blob/dev/common/document_parser/ref_utils.py) creates a dictionary of reference types and regex patterns. 

_Originally_, each value in the dictionary was a tuple of patterns, where the first pattern was identical to the ending part of the second pattern. 
Ex:
```
(
    re.compile(r"(([A-Z]+-)?[0-9]{4}\.\s*[0-9]{1,3}\s*(-[A-Z]+)?([E])?)", re.IGNORECASE),
    re.compile(r"\b(((dod\s*placeholder)|(dod))\s*\)?\s*([A-Z]+-)?[0-9]{4}\.\s*[0-9]{1,3}\s*(-[A-Z]+)?([E])?)", re.IGNORECASE)
)
```

In [`look_for_general()`](https://github.com/dod-advana/gamechanger-data/blob/dev/common/document_parser/lib/ref_list.py), the second pattern (full reference) was searched for, and then the first pattern (numerical part of a reference) was searched for within the second pattern.

So, **problem number 1** is duplicated regex search.
**Problem number 2** is the patterns contain many (unnecessary) capture groups, and that slows performance. ([reference](https://stackoverflow.com/questions/41444807/why-is-regex-search-slower-with-capturing-groups-in-python))
Both of these problems can be **solved by minimizing the number of capture groups** in the patterns.

_The changes in this PR_ make it so that each pattern has exactly 1 capture group that is the numerical part of a reference.
   

## Time difference:
![reference_extraction_time_improvement](https://user-images.githubusercontent.com/105378552/181646095-b0ce7271-7f59-4544-8cf4-09187d6e9bae.png)

## How to Test:
Run test functions in [test_ref.py](https://github.com/dod-advana/gamechanger-data/blob/parser-efficiency-references/common/document_parser/test_ref.py).
- See [here](https://stackoverflow.com/questions/28643534/is-there-a-way-in-python-to-execute-all-functions-in-a-file-without-explicitly-c) for how to easily run all functions in a file (note: skip over `check()`, it is a helper function for the tests).
